File: 01_introduction_to_algorithms/javascript/01_binary_search.js
'use strict';

function binarySearch(sortedList, item) {
  var low = 0;
  var high = sortedList.length - 1;
  
  while (low <= high) {
    var middle = Math.floor((low + high) / 2);
    var guess = sortedList[middle];

    if (guess === item) {
      return middle;
    }
    if (guess > item) {
      high = middle - 1;
    } else {
      low = middle + 1;
    }
  }
  
  return null;
}

var my_list = [1, 3, 5, 7, 9];

console.log(binarySearch(my_list, 3)); // 1
console.log(binarySearch(my_list, -1)); // null

File: 01_introduction_to_algorithms/c/01_binary_search.c
#include <stdio.h>
#include <stdlib.h>

#define SIZE 4;

int binarySearch(int[], int);

int main()
{
    int myList[] = {1, 3, 5, 7, 9};
    printf("%d\n", binarySearch(myList, 3));  // 1
    printf("%d\n", binarySearch(myList, -1)); //-1
    return 0;
}

int binarySearch(int list[], int item)
{
    int low = 0;
    int high = SIZE;
    while (low <= high)
    {
        int mid = (int)((low + high) / 2);
        int guess = list[mid];

        if (guess == item)
        {
            return mid;
        }
        else if (guess > item)
        {
            high = mid - 1;
        }
        else
        {
            low = mid + 1;
        }
    }
    return -1; //number not find
}
